text/0000-rand-core.md

```diff
@@ -0,0 +1,1099 @@
+type. Probably either is fine.
+
+Existing `Seed` types used by implementations are `[u32; 4]`, `&[u32]` and
+`&[u64]`. The intention of using slices is to allow partial seeding; it is
+perfectly valid for these PRNGs to have some (or even all) of their state seeded
+to 0. The disadvantages of this are that seed types are inconsistent between
+generators and are not `Sized`, so cannot be generated by [`Rand`] or similar
+without also specifying a size. We could instead suggest implementation only
+for `Sized` types, or for a `Sized` type *and* a byte-slice (`&[u8]`).
+
+We could move the `Seed` type from a trait parameter to an associated type:
+
+```rust
+pub trait SeedableRng: Rng {
+    type Seed;
+    ..
+}
+```
+
+Personally I would rather not do this: it prevents implementation of
+`SeedableRng` for multiple types.
+
+Instead of using a generic `Seed` type, we could have multiple versions of
+`SeedableRng` each with their own fixed seed type, e.g. one for `u64` or
+`(u64, u64)` for simple specification (optionally with stream support), and
+another for `&[u8]` allowing seeding from arbitrary byte slices (possibly with
+a function `fn seed_len() -> u32` to specify the ideal seed length in bytes).
+We could do this in addition to a generic version of `SeedableRng` taking the
+full seed as a `Sized` type, or even without this.
+
+### Creation of securely-seeded RNGs
+
+Often, PRNGs are seeded from "somewhat random" sources such as the system
+clock. Rust's `rand` tries to make the best option easy by making it possible
+to construct random number generators from the system generator, e.g. the
+[`ChaChaRng`] type supports the [`Rand`] trait. This makes it possible to
+construct a securely seeded `ChaChaRng` with:
+
+```rust
+use rand::{Rng, OsRng, ChaChaRng};
+
+// OsRng::new() returns a Result
+let mut rng: ChaChaRng = OsRng::new().unwrap().gen();
+```
+
+This RFC seeks to introduce an alternative:
+
+```rust
+// items may be moved to other crates, but for now at least are accessible here:
+use rand::{ChaChaRng, NewSeeded};
+
+// new() returns a Result
+let mut rng = ChaChaRng::new().unwrap();
+```
+
+Here, `NewSeeded` is a trait providing the `new` function. It is automatically
+implemented for any type implementing `SeedFromRng`:
+
+```rust
+/// Support mechanism for creating securely seeded objects 
+/// using the OS generator.
+/// Intended for use by RNGs, but not restricted to these.
+/// 
+/// This is implemented automatically for any PRNG implementing `SeedFromRng`,
+/// and for normal types shouldn't be implemented directly. For mock generators
+/// it may be useful to implement this instead of `SeedFromRng`.
+#[cfg(feature="std")]
+pub trait NewSeeded: Sized {
+    /// Creates a new instance, automatically seeded via `OsRng`.
+    fn new() -> Result<Self>;
+}
+
+#[cfg(feature="std")]
+impl<R: SeedFromRng> NewSeeded for R {
+    fn new() -> Result<Self> {
+        let mut r = OsRng::new()?;
+        Self::from_rng(&mut r)
+    }
+}
```

Comment:
Such implementation essentially makes it impossible to implement `NewSeeded` by anybody else for any other seedable RNG.

That is, it is impossible to write something along the lines of

```rust
struct DifferentlySeededRng(MyRng);
impl SeedFromRng for DifferentlySeededRng {...}
impl NewSeeded for DifferentlySeededRng {
    fn new() -> Result<Self> { ... }
}
```

due to coherence, which makes this trait hardly useful IMO. If we’re adding a trait+impl like this, we should aim to make `fn new` a `default fn new` as soon as specialisation is stable… but I personally think something like this:

```rust
impl OsRng {
    fn seed_new<R: SeedFromRng>() -> Result<R>;
}
```

is potentially way better and is a pattern that’s extensible to any number of seed-capable RNGs.

Reply:
Alternatively a plain function could exist in `rand` without any of that unnecessary trait stuff.

Reply:
The trait I suggested is there to allow this:

```rust
use NewSeeded;
let rng = SomePRNG::new();
```

I suppose `let rng = OsRng::seed_new::<SomePRNG>();` isn't a *terrible* alternative.

To be honest, I don't see *why* anyone would need to write an alternative implementation for `NewSeeded` though.

Reply:
Follow-up comments [in this issue](https://github.com/dhardy/rand/issues/17) please.